pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1633 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nftables
+
+//
+// NOTE: this needs to be tested in e2e since it uses nftables for everything.
+//
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base32"
+	"fmt"
+	"net"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/danwinship/knftables"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/events"
+	utilsysctl "k8s.io/component-helpers/node/util/sysctl"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/proxy"
+	"k8s.io/kubernetes/pkg/proxy/conntrack"
+	"k8s.io/kubernetes/pkg/proxy/healthcheck"
+	"k8s.io/kubernetes/pkg/proxy/metaproxier"
+	"k8s.io/kubernetes/pkg/proxy/metrics"
+	proxyutil "k8s.io/kubernetes/pkg/proxy/util"
+	proxyutiliptables "k8s.io/kubernetes/pkg/proxy/util/iptables"
+	"k8s.io/kubernetes/pkg/util/async"
+	utilexec "k8s.io/utils/exec"
+	netutils "k8s.io/utils/net"
+	"k8s.io/utils/ptr"
+)
+
+const (
+	// the nftables table
+	kubeProxyTable = "kube-proxy"
+
+	// service dispatch
+	kubeServicesChain       = "services"
+	kubeServiceIPsMap       = "service-ips"
+	kubeServiceNodePortsMap = "service-nodeports"
+
+	// set of IPs that accept NodePort traffic
+	kubeNodePortIPsSet = "nodeport-ips"
+
+	// handling for services with no endpoints
+	kubeEndpointsCheckChain    = "endpoints-check"
+	kubeNoEndpointServicesMap  = "no-endpoint-services"
+	kubeNoEndpointNodePortsMap = "no-endpoint-nodeports"
+	kubeRejectChain            = "reject-chain"
+
+	// LoadBalancerSourceRanges handling
+	kubeFirewallSet             = "firewall"
+	kubeFirewallCheckChain      = "firewall-check"
+	kubeFirewallAllowSet        = "firewall-allow"
+	kubeFirewallAllowCheckChain = "firewall-allow-check"
+
+	// masquerading
+	kubeMarkMasqChain     = "mark-for-masquerade"
+	kubeMasqueradingChain = "masquerading"
+
+	// chain for special filtering rules
+	kubeForwardChain = "forward"
+)
+
+const sysctlNFConntrackTCPBeLiberal = "net/netfilter/nf_conntrack_tcp_be_liberal"
```

Comment:
Pretty much everything in `pkg/proxy` could benefit from a bit of refactoring...

For now this is just trying to make an nftables backend that looks like the iptables backend and I was mostly trying to avoid additional refactoring in this PR.